zebra-state/src/service/read/difficulty.rs

```diff
@@ -0,0 +1,64 @@
+//! Get context and calculate difficulty for the next block.
+
+use std::borrow::Borrow;
+
+use zebra_chain::{
+    block::{Block, Hash, Height},
+    parameters::{Network, POW_AVERAGING_WINDOW},
+    work::difficulty::CompactDifficulty,
+};
+
+use crate::service::{
+    any_ancestor_blocks,
+    check::{difficulty::POW_MEDIAN_BLOCK_SPAN, AdjustedDifficulty},
+    finalized_state::ZebraDb,
+    NonFinalizedState,
+};
+
+/// Return the CompactDifficulty for the current best chain.
+///
+/// Return `None` if we don't have enough blocks in the state.
+pub fn relevant_chain_difficulty(
+    non_finalized_state: &NonFinalizedState,
+    db: &ZebraDb,
+    tip: (Height, Hash),
+    network: Network,
+) -> Option<CompactDifficulty> {
+    let relevant_chain = any_ancestor_blocks(non_finalized_state, db, tip.1);
+    difficulty(relevant_chain, tip.0, network)
+}
+
+fn difficulty<C>(
+    relevant_chain: C,
+    tip_height: Height,
+    network: Network,
+) -> Option<CompactDifficulty>
+where
+    C: IntoIterator,
+    C::Item: Borrow<Block>,
+    C::IntoIter: ExactSizeIterator,
+{
+    const MAX_CONTEXT_BLOCKS: usize = POW_AVERAGING_WINDOW + POW_MEDIAN_BLOCK_SPAN;
+
+    let relevant_chain: Vec<_> = relevant_chain
+        .into_iter()
+        .take(MAX_CONTEXT_BLOCKS)
+        .collect();
+
+    let relevant_data = relevant_chain.iter().map(|block| {
+        (
+            block.borrow().header.difficulty_threshold,
+            block.borrow().header.time,
+        )
+    });
+
+    if relevant_data.len() < MAX_CONTEXT_BLOCKS {
+        return None;
+    }
```

Comment:
Can we turn this into an `expect()` or returning an error?
Or do we need it for the tests?

```suggestion
    // The getblocktemplate RPC returns an error if Zebra is not synced to the tip.
    // So this will never happen in production code.
    if relevant_data.len() < MAX_CONTEXT_BLOCKS {
        return None;
    }
```

Reply:
I think this is fine, we can have a panic here but what is the big deal with this as it is now ?

Reply:
It makes the request slightly harder to use, but it is ok to merge.